Use system context for cluster state update tasks

ywelsch opened the pull request:
This PR makes it so that cluster state update tasks always run under the system context, only restoring the original context when the listener that was provided with the task is called. A notable exception is the `clusterStatePublished(...)` callback which will still run under system context, because it's defined on the executor-level, and not the task level, and only called once for the combined batch of tasks and can therefore not be uniquely identified with a task / thread context. I've looked at usages of this, and adapted the two problematic ones to use `clusterStateProcessed` instead.

Relates #30603

elasticmachine commented:
Pinging @elastic/es-distributed

bleskes commented on server/src/main/java/org/elasticsearch/cluster/service/MasterService.java:
```diff
@@ -710,9 +715,13 @@ public void onNoLongerMaster() {
+        final ThreadContext threadContext = threadPool.getThreadContext();
+        final Supplier<ThreadContext.StoredContext> supplier = threadContext.newRestorableContext(false);
+        try (ThreadContext.StoredContext ignore = threadContext.stashContext()) {
+            threadContext.markAsSystemContext();
```
instead of doing this, I wonder if we should invest in a SystemThreadContext that we give to the executor and it does is under the system context. Alternatively we can have a "no context" executor that simply doesn't set any context. @tvernum do you see any other usage for this?

ywelsch replied:
I'm not sure if I understand this correctly. Would you have different executors based on whether they switch to the system context or stay in default context? Looking at the other usages of `threadContext.markAsSystemContext();`, all of them seem to use executors that cannot be categorized as system or non-system (i.e. generic threadpool, ...).

bleskes replied:
my suggestion referred to the entire pattern of generating a new context and then marking it as system and deal with on the executor level. This has two potential upsides:
1) You don't need to remember to do this everywhere you submit a task (not so important here as there's just one place).
2) We don't do all the "clean & capture and then restore in the thread pool" song and dance.

Since this is not a high throughput usage and we only have one place we submit tasks it doesn't have such a big impact on `MasterService` but might be useful for other things that use an internal thread pool executor.

bleskes commented on x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/job/JobManager.java:
```diff
@@ -346,8 +346,8 @@ public void onFailure(String source, Exception e) {
                 }
 
                 @Override
-                public void clusterStatePublished(ClusterChangedEvent clusterChangedEvent) {
```
I wonder if ClusterStateUpdateTask should have a final empty method for this? I think it's always the wrong one in that contex.t

ywelsch replied:
I like this idea, I've pushed 1b7eeb4 to implement this.

ywelsch commented:
The doc tests (`:docs:integTestRunner`) are currently failing on this, as there's an assertion in `RemoteClusterConnection` that checks that `SniffClusterStateResponseHandler.handleResponse` is not called in a system context. I'm not sure why that assertion is there, or what's triggering the failure. Maybe @s1monw can help with this?

Finally, I'm not sure what the implications of this PR are for cluster state publication, where we cross node boundaries (which does not seem to transfer the system context, only reestablish it because of internal action?)

ywelsch commented:
I've removed the assertion in `RemoteClusterConnection`, which no longer holds. The reason is that `forkConnect` is called from the cluster applier thread, which now runs under system context.
To validate this, I added an assertion in 7b92bdb and got the following stack trace:

```
[2018-06-13T09:35:44,461][ERROR][o.e.b.ElasticsearchUncaughtExceptionHandler] [node-0] fatal error in thread [elasticsearch[node-0][clusterApplierService#updateTask][T#1]], exiting
java.lang.AssertionError: context is a system context
        at org.elasticsearch.transport.RemoteClusterConnection$ConnectHandler.forkConnect(RemoteClusterConnection.java:377) ~[elasticsearch-7.0.0-alpha1-SNAPSHOT.jar:7.0.0-alpha1-SNAPSHOT]
        at org.elasticsearch.transport.RemoteClusterConnection$ConnectHandler.connect(RemoteClusterConnection.java:371) ~[elasticsearch-7.0.0-alpha1-SNAPSHOT.jar:7.0.0-alpha1-SNAPSHOT]
        at org.elasticsearch.transport.RemoteClusterConnection$ConnectHandler.connect(RemoteClusterConnection.java:335) ~[elasticsearch-7.0.0-alpha1-SNAPSHOT.jar:7.0.0-alpha1-SNAPSHOT]
        at org.elasticsearch.transport.RemoteClusterConnection.updateSeedNodes(RemoteClusterConnection.java:135) ~[elasticsearch-7.0.0-alpha1-SNAPSHOT.jar:7.0.0-alpha1-SNAPSHOT]
        at org.elasticsearch.transport.RemoteClusterService.updateRemoteClusters(RemoteClusterService.java:148) ~[elasticsearch-7.0.0-alpha1-SNAPSHOT.jar:7.0.0-alpha1-SNAPSHOT]
        at org.elasticsearch.transport.RemoteClusterService.updateRemoteCluster(RemoteClusterService.java:327) ~[elasticsearch-7.0.0-alpha1-SNAPSHOT.jar:7.0.0-alpha1-SNAPSHOT]
        at org.elasticsearch.transport.RemoteClusterService.updateRemoteCluster(RemoteClusterService.java:314) ~[elasticsearch-7.0.0-alpha1-SNAPSHOT.jar:7.0.0-alpha1-SNAPSHOT]
        at org.elasticsearch.common.settings.Setting$AffixSetting$1.lambda$getValue$0(Setting.java:626) ~[elasticsearch-7.0.0-alpha1-SNAPSHOT.jar:7.0.0-alpha1-SNAPSHOT]
        at org.elasticsearch.common.settings.Setting$Updater.apply(Setting.java:957) ~[elasticsearch-7.0.0-alpha1-SNAPSHOT.jar:7.0.0-alpha1-SNAPSHOT]
        at org.elasticsearch.common.settings.Setting$AffixSetting$1.apply(Setting.java:641) ~[elasticsearch-7.0.0-alpha1-SNAPSHOT.jar:7.0.0-alpha1-SNAPSHOT]
        at org.elasticsearch.common.settings.Setting$AffixSetting$1.apply(Setting.java:611) ~[elasticsearch-7.0.0-alpha1-SNAPSHOT.jar:7.0.0-alpha1-SNAPSHOT]
        at org.elasticsearch.common.settings.AbstractScopedSettings$SettingUpdater.lambda$updater$0(AbstractScopedSettings.java:410) ~[elasticsearch-7.0.0-alpha1-SNAPSHOT.jar:7.0.0-alpha1-SNAPSHOT]
        at org.elasticsearch.common.settings.AbstractScopedSettings.applySettings(AbstractScopedSettings.java:170) ~[elasticsearch-7.0.0-alpha1-SNAPSHOT.jar:7.0.0-alpha1-SNAPSHOT]
        at org.elasticsearch.cluster.service.ClusterApplierService.applyChanges(ClusterApplierService.java:472) ~[elasticsearch-7.0.0-alpha1-SNAPSHOT.jar:7.0.0-alpha1-SNAPSHOT]
        at org.elasticsearch.cluster.service.ClusterApplierService.runTask(ClusterApplierService.java:430) ~[elasticsearch-7.0.0-alpha1-SNAPSHOT.jar:7.0.0-alpha1-SNAPSHOT]
        at org.elasticsearch.cluster.service.ClusterApplierService$UpdateTask.run(ClusterApplierService.java:160) ~[elasticsearch-7.0.0-alpha1-SNAPSHOT.jar:7.0.0-alpha1-SNAPSHOT]
        at org.elasticsearch.common.util.concurrent.ThreadContext$ContextPreservingRunnable.run(ThreadContext.java:625) ~[elasticsearch-7.0.0-alpha1-SNAPSHOT.jar:7.0.0-alpha1-SNAPSHOT]
        at org.elasticsearch.common.util.concurrent.PrioritizedEsThreadPoolExecutor$TieBreakingPrioritizedRunnable.runAndClean(PrioritizedEsThreadPoolExecutor.java:244) ~[elasticsearch-7.0.0-alpha1-SNAPSHOT.jar:7.0.0-alpha1-SNAPSHOT]
        at org.elasticsearch.common.util.concurrent.PrioritizedEsThreadPoolExecutor$TieBreakingPrioritizedRunnable.run(PrioritizedEsThreadPoolExecutor.java:207) ~[elasticsearch-7.0.0-alpha1-SNAPSHOT.jar:7.0.0-alpha1-SNAPSHOT]
        at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1135) ~[?:?]
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:635) ~[?:?]
        at java.lang.Thread.run(Thread.java:844) [?:?]
```

s1monw commented:
@ywelsch I think you can remove the assertion it's not valid anymore. The new assertion also isn't

ywelsch commented:
@s1monw yes, thanks, I only added that extra assertion to see what was causing the existing assertion to trip. I have removed those assertions in faf152c.

ywelsch commented:
After removing the assertions in `RemoteClusterConnection`, I ran into an ML test failure (`testLookbackWithoutPermissions`). @davidkyle has looked into this and pushed a fix to this PR: 8b8e3e3. ML was previously relying on the behavior that the cluster state update task (the `execute` method) was run as the user that submitted the task (as it stored the headers of the originating request into the cluster state, in order to run the datafeeds with the permission of the last editing user). This was changed by this PR and the fix is therefore to explicitly pass the desired information to the task (see 8b8e3e3).

ywelsch commented:
@tvernum ping :)

ywelsch commented:
Thanks @bleskes @tvernum @davidkyle @s1monw
